pkg/backup/backuper.go

```diff
@@ -0,0 +1,321 @@
+package backup
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"os"
+	"path"
+	"time"
+
+	"github.com/mattn/go-sqlite3"
+
+	"github.com/pkg/errors"
+)
+
+// BackupFilenamePrefix is the prefix used in every backup file.
+const BackupFilenamePrefix = "tbl_backup"
+
+var (
+	open          = openDatabase
+	newBackupFile = createBackupFile
+)
+
+// Backuper is the process that executes the backup process.
+type Backuper struct {
+	source DB
+	result *BackupResult
+	config *Config
+}
+
+// NewBackuper creates a new backuper responsible for making backups of a SQLite database.
+func NewBackuper(sourcePath string, backupDir string, opts ...Option) (*Backuper, error) {
+	config := DefaultConfig()
+	for _, o := range opts {
+		if err := o(config); err != nil {
+			return nil, err
+		}
+	}
+
+	b := &Backuper{
+		result: &BackupResult{
+			Dir: backupDir,
+		},
+		config: config,
+	}
+
+	if err := b.init(sourcePath, backupDir); err != nil {
+		return nil, err
+	}
+
+	return b, nil
+}
+
+// Backup creates a backup to a temporary in disk file.
+func (b *Backuper) Backup(ctx context.Context) (*BackupResult, error) {
+	// SQLite backup starts
+	startTime := time.Now()
+
+	connA, err := b.source.Conn(ctx)
+	if err != nil {
+		return b.result, errors.Errorf("getting db conn: %s", err)
+	}
+	defer func() {
+		_ = connA.Close()
+	}()
+
+	connB, err := b.result.Database.Conn(ctx)
+	if err != nil {
+		return b.result, errors.Errorf("getting backup db conn: %s", err)
+	}
+	defer func() {
+		_ = connB.Close()
+	}()
+
+	if err := b.doBackup(connA, connB); err != nil {
+		return b.result, errors.Errorf("backup: %s", err)
+	}
+
+	// SQLite backup finishes
+	b.result.ElapsedTime = time.Since(startTime)
+
+	// count the number of tables for track record
+	tableCount, err := b.doTableCount(ctx, connB)
+	if err != nil {
+		return b.result, errors.Errorf("table count: %s", err)
+	}
+
+	// gets the size of backup for track record
+	backupSize, err := b.getFileSize(b.result.Database.Path())
+	if err != nil {
+		return b.result, errors.Errorf("get file size: %s", err)
+	}
+
+	if b.config.Vacuum {
+		b.result.SizeAfterVacuum, b.result.VacuumElapsedTime, err = b.doVacuum(ctx, connB)
+		if err != nil {
+			return b.result, errors.Errorf("vacuum: %s", err)
+		}
+	}
+
+	if b.config.Compression {
+		panic("compression not implemented")
+	}
+
+	if b.config.Pruning {
+		panic("pruning not implemented")
+	}
+
+	b.result.SizeBeforeVacuum = backupSize
+	b.result.TableCount = tableCount
+	return b.result, nil
+}
+
+// init opens databases and ping them, then initializes variables.
+func (b *Backuper) init(sourcePath string, backupDir string) error {
+	source, err := open(sourcePath)
+	if err != nil {
+		return errors.Errorf("opening source db: %s", err)
+	}
+
+	filename, timestamp, err := newBackupFile(backupDir, time.Now().UTC())
```

Comment:
Check [createBackupFile](https://github.com/tablelandnetwork/go-tableland/pull/227/files#diff-c95ef053206f27ba2307970f3b1bfaa55fcd806df5e69e44123c010ccf796ac7R227). There's a mocked version of this in tests as well.

Reply:
If you jumped to here, go back to [Backup method](https://github.com/tablelandnetwork/go-tableland/pull/227/files#diff-c95ef053206f27ba2307970f3b1bfaa55fcd806df5e69e44123c010ccf796ac7R55).